File: backend-js/src/shared/relay/cursorCodec.ts
import { DomainError } from '../errors/DomainError';

export interface DecodedCursor {
  createdAt: Date;
  id: string;
}

export function encodeCursor(createdAt: Date, id: string): string {
  return Buffer.from(`${createdAt.getTime()}:${id}`, 'utf8').toString('base64url');
}

export function decodeCursor(cursor: string): DecodedCursor {
  try {
    const decoded = Buffer.from(cursor, 'base64url').toString('utf8');
    const separator = decoded.indexOf(':');
    if (separator <= 0 || separator === decoded.length - 1) {
      throw DomainError.validation('Invalid cursor format');
    }

    const timestamp = Number(decoded.slice(0, separator));
    const id = decoded.slice(separator + 1);

    if (!Number.isFinite(timestamp)) {
      throw DomainError.validation('Invalid cursor format');
    }

    return {
      createdAt: new Date(timestamp),
      id,
    };
  } catch (error) {
    if (error instanceof DomainError) {
      throw error;
    }
    throw DomainError.validation('Invalid cursor encoding');
  }
}

File: backend-java/src/main/java/com/woovi/crudbank/shared/relay/CursorCodec.java
package com.woovi.crudbank.shared.relay;

import com.woovi.crudbank.shared.error.DomainException;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Base64;

@Component
public class CursorCodec {

    public String encode(Instant createdAt, String id) {
        String raw = createdAt.toEpochMilli() + ":" + id;
        return Base64.getUrlEncoder().encodeToString(raw.getBytes(StandardCharsets.UTF_8));
    }

    public DecodedCursor decode(String cursor) {
        try {
            String decoded = new String(Base64.getUrlDecoder().decode(cursor), StandardCharsets.UTF_8);
            int separator = decoded.indexOf(':');
            if (separator <= 0 || separator == decoded.length() - 1) {
                throw DomainException.validation("Invalid cursor format");
            }
            long timestamp = Long.parseLong(decoded.substring(0, separator));
            String id = decoded.substring(separator + 1);
            return new DecodedCursor(Instant.ofEpochMilli(timestamp), id);
        } catch (IllegalArgumentException ex) {
            throw DomainException.validation("Invalid cursor encoding");
        }
    }

    public record DecodedCursor(Instant createdAt, String id) {
    }
}
